### PR TITLE
refactor decode logic, add file decoder.scala

### PR DESCRIPTION
**Pull request by AlwenXXD**
**Problem Summary**: I think the readability of decode logic is a little bit poor, and hard to extend instructions.

**What's Changed**: Now most of contorl signal is defined in a table in file Decoder.scala using ChiselEnum.

**How it Works**: When add new modules or new instructions, you can esaily add instruction decode information and control signals in decode table.

### CODE REVIEW

File: src/main/scala/ALU.scala [hunk 1]
```diff
@@ -3,58 +3,58 @@ import chisel3.util._
 
 class ALU extends Module {
 
-  import ALUOperation._
+//  import ALUOperation._
```
Comment (longfangsong) [inline src/main/scala/ALU.scala:6]: Feel free to delete the code if you think they are not useful.

**Review by zhoudian64 — APPROVED** [review]
LGTM, what's your opinion. @longfangsong

File: src/main/scala/CPU.scala [hunk 2]
```diff
@@ -34,52 +37,49 @@ class CPU extends Module {
...
     stall := false.B
     regFile.io.writeEnable := true.B
-    io.dataBusBundle.address := (regFile.io.outputA.asSInt() + immGen.io.result).asUInt()
-    io.dataBusBundle.maskLevel := instruction(13, 12)
+    io.dataBusBundle.address := (regFile.io.outputA.asSInt() + decoder.imm_data.asSInt()).asUInt()
+    io.dataBusBundle.maskLevel := decoder.maskLevel
     // second part of load
-    switch(instruction(14, 12)) {
+    switch(decoder.uop) {
       // LB
```
Comment (longfangsong) [inline src/main/scala/CPU.scala:65]: You can remove the comments here if you use named `ChiselEnum`.

File: src/main/scala/Decoder.scala [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    EBREAK     -> List(uOP.EBREAK    , InstType.system   , ImmSel.no_imm  ),
+    MRET       -> List(uOP.MRET      , InstType.system   , ImmSel.no_imm  ),
+    SRET       -> List(uOP.SRET      , InstType.system   , ImmSel.no_imm  ),
+    SFENCE_VMA -> List(uOP.SFENCE_VMA, InstType.fence    , ImmSel.no_imm  ),
+    WFI        -> List(uOP.WFI       , InstType.system   , ImmSel.no_imm  ),
+    )
+
+}
+
+object uOP extends ChiselEnum {
```
Comment (longfangsong) [inline src/main/scala/Decoder.scala:171]: I don't know whether `uOP` is the best name here...
However I don't have a better alternative.
Please at least make it `UOp`.

File: src/main/scala/Decoder.scala [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+  SUB,
+  SLL,
+  SLT,
+  SLTU,
+  XOR,
+  SRL,
+  SRA,
+  OR,
+  AND,
+  //load,store
```
Comment (longfangsong) [inline src/main/scala/Decoder.scala:196]:
```suggestion
  //load, store
```

File: src/main/scala/Decoder.scala [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+class Decoder extends Bundle {
+  val uop      = uOP()
+  val inst_type = InstType()
+  val need_imm = Bool()
+  val rs1_addr = UInt(5.W)
+  val rs2_addr = UInt(5.W)
+  val rd_addr  = UInt(5.W)
+  val imm_data = UInt(32.W)
+  val csr_idx  = UInt(12.W)
```
Comment (longfangsong) [inline src/main/scala/Decoder.scala:14]: We prefer `camelCase` for names.

**Review by longfangsong — CHANGES_REQUESTED** [review]
Rest LGTM

File: src/main/scala/Decoder.scala [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+      ImmSel.is_zimm.asUInt() -> zimm,
+    ).toSeq)
+    maskLevel := inst(13, 12)
+    csr_idx := inst(31, 20)
+    this
+  }
+}
+
+object Instructions {
+  def BEQ                = BitPat("b?????????????????000?????1100011")
```
Comment (longfangsong) [inline src/main/scala/Decoder.scala:57]: We prefer `camelCase` for method names

**Review by longfangsong — APPROVED** [review]
LGTM, thank you for your contribution!

**Pull request merged**